Add feature gate ExpandedDNSConfig

gjkim42 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

This PR implements [KEP 2595 (Expanded DNS Configuration)](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2595-expanded-dns-config)

Add feature gate `ExpandedDNSConfig` that enables apiserver and kubelet to allow up to 32 DNSSearchPaths and up to 2048 DNSSearchListChars.

Original issue: #100582
KEP issue: https://github.com/kubernetes/enhancements/issues/2595
ref: https://github.com/kubernetes/kubernetes/issues/100622

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
New feature gate "ExpandedDNSConfig" is now available. This feature allows Kubernetes to have expanded DNS configuration.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
[KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2595-expanded-dns-config
```

gjkim42 commented:
/hold

gjkim42 commented:
/cc @aojea @liggitt @thockin 
Regarding https://github.com/kubernetes/kubernetes/pull/100583#discussion_r603308918,

Made an alternative to #100583 to compare.
If you prefer this way, let me know.

This PR enables apiserver and kubelet to allow ExpandedDNSConfig at once.

thockin commented on pkg/kubelet/network/dns/dns.go:
```diff
@@ -99,19 +102,31 @@ func omitDuplicates(strs []string) []string {
 	limitsExceeded := false
 
-	if len(composedSearch) > validation.MaxDNSSearchPaths {
-		composedSearch = composedSearch[:validation.MaxDNSSearchPaths]
```
I think you should still keep something like this - before you iterate aan unbounded list - but truncate to the new, larger value and flag limitsExceeded here

thockin commented on pkg/kubelet/network/dns/dns.go:
```diff
@@ -99,19 +102,31 @@ func omitDuplicates(strs []string) []string {
-		composedSearch = composedSearch[:validation.MaxDNSSearchPaths]
+	for i, search := range composedSearch {
+		if len(search) > utilvalidation.DNS1123SubdomainMaxLength {
+			composedSearch = composedSearch[:i]
```
In this case I think it's safer to leave it in, invalid or not.  Truncating is pretty hostile.

aojea replied:
is not better to just skip only the ones that exceeds the limits and don't break?
if I have a search with"
s1 badSearch300chars s2 s3 s4

I will have s1 s2 s3 s4, with this I'll have s1 IIUIRC

aojea replied:
https://github.com/golang/go/wiki/SliceTricks#filter-in-place

gjkim42 replied:
Done.

thockin commented on pkg/kubelet/network/dns/dns_test.go:
```diff
@@ -121,43 +125,62 @@ func TestFormDNSSearchFitsLimits(t *testing.T) {
+			hostNames:              []string{"testNS.svc.TEST", "svc.TEST", "TEST", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "23", "24", "25", "26", "27", "28", "29", "30", "31", "32", "33"},
+			resultSearch:           []string{"testNS.svc.TEST", "svc.TEST", "TEST", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "23", "24", "25", "26", "27", "28", "29", "30", "31", "32"},
+			events:                 []string{"Search Line limits were exceeded, some search paths have been omitted, the applied search line is: testNS.svc.TEST svc.TEST TEST 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32"},
+			allowExpandedDNSConfig: true,
```
How about a success case with this set and a long line

gjkim42 replied:
Done.

thockin commented on pkg/kubelet/network/dns/dns_test.go:
```diff
@@ -445,8 +468,14 @@ func TestGetPodDNS(t *testing.T) {
-		expLength = 6
+
+	maxDNSSearchPaths := validation.MaxDNSSearchPathsLegacy
+	if utilfeature.DefaultFeatureGate.Enabled(features.ExpandedDNSConfig) {
```
You can't use the gate like this in a test - you need to run the test cases both with and without it

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3325,7 +3338,7 @@ func ValidatePodSpec(spec *core.PodSpec, podMeta *metav1.ObjectMeta, fldPath *fi
 	allErrs = append(allErrs, validateImagePullSecrets(spec.ImagePullSecrets, fldPath.Child("imagePullSecrets"))...)
 	allErrs = append(allErrs, validateAffinity(spec.Affinity, fldPath.Child("affinity"))...)
-	allErrs = append(allErrs, validatePodDNSConfig(spec.DNSConfig, &spec.DNSPolicy, fldPath.Child("dnsConfig"))...)
+	allErrs = append(allErrs, validatePodDNSConfig(spec.DNSConfig, &spec.DNSPolicy, fldPath.Child("dnsConfig"), opts.AllowExpandedDNSConfig)...)
```
I don't see where this option is ever set?

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -3002,12 +3008,16 @@ func validatePodDNSConfig(dnsConfig *core.PodDNSConfig, dnsPolicy *core.DNSPolic
-		if len(dnsConfig.Searches) > MaxDNSSearchPaths {
-			allErrs = append(allErrs, field.Invalid(fldPath.Child("searches"), dnsConfig.Searches, fmt.Sprintf("must not have more than %v search paths", MaxDNSSearchPaths)))
+		maxDNSSearchPaths, maxDNSSearchListChars := MaxDNSSearchPathsLegacy, MaxDNSSearchListCharsLegacy
+		if allowExpandedDNSConfig {
```
This is not right for update.  You need to check if the old spec already had long data AND the new spec is the same, in which case you validate using the longer limits.  If the old spec was short and the new spec is long, that's an error.  Also, if the old spec was long and the new spec is long, that's an error.

Kubelet always respects the gate and truncates if needed.

gjkim42 commented:
/retest

gjkim42 reviewed (commented):
Updated.

gjkim42 commented:
KEP is now merged.
Updated the description.

gjkim42 commented:
/retest

gjkim42 commented:
/retest

gjkim42 commented:
/retest

aojea commented:
> @gjkim42: The following test **failed**, say `/retest` to rerun all failed tests:
> 
> Test name	Commit	Details	Rerun command
> pull-kubernetes-e2e-gce-alpha-features	[89d698c](https://github.com/kubernetes/kubernetes/commit/89d698c79c895c53b11aea3f86654b37aaa34be2)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/100651/pull-kubernetes-e2e-gce-alpha-features/1394300091778469888/)	`/test pull-kubernetes-e2e-gce-alpha-features`
> [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes&repo=kubernetes&pr=100651). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3Agjkim42). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes/kubernetes/issues?q=is:issue+is:open) when you hit one in your PR.

the job is failing because kubelet can't start, it complains about

```
E0517 14:57:53.136220    2576 server.go:218] "Failed to set feature gates from initial flags-based config" err="unrecognized feature gate: DisableCloudProviders"
kubelet.service: Main process exited, code=exited, status=1/FAILURE
kubelet.service: Failed with result 'exit-code'.
```

@liggitt  @BenTheElder  does it ring a bell?

liggitt commented:
yes, test-infra config was changed (https://github.com/kubernetes/test-infra/pull/22035) before the feature gate was merged (https://github.com/kubernetes/kubernetes/pull/100136, still not merged)

gjkim42 commented:
Thanks. I'll wait for #100136 to be merged.

BenTheElder commented:
Note that `pull-kubernetes-e2e-gce-alpha-features` is not strictly required to merge.

gjkim42 commented:
> Note that `pull-kubernetes-e2e-gce-alpha-features` is not strictly required to merge.

Thanks @BenTheElder 

---

ping @thockin @aojea @liggitt 
Could you PTAL?

ehashman commented:
/priority important-soon
/triage accepted

Not sure that this needs Node reviewers, seems like a network thing.

aojea commented:
> Not sure that this needs Node reviewers, seems like a network thing.

it is dns ... the more the merrier XD

aojea commented on pkg/kubelet/network/dns/dns.go:
```diff
@@ -99,19 +102,31 @@ func omitDuplicates(strs []string) []string {
+	if utilfeature.DefaultFeatureGate.Enabled(features.ExpandedDNSConfig) {
+		maxDNSSearchPaths, maxDNSSearchListChars = validation.MaxDNSSearchPathsExpanded, validation.MaxDNSSearchListCharsExpanded
+	}
+	if len(composedSearch) > maxDNSSearchPaths {
```
I understand Tim's point and I think he is right, if we truncate here the for in https://github.com/kubernetes/kubernetes/pull/100651/files#r607368307 doesn't have to go through searches that may be truncated later

gjkim42 replied:
I've got your point. Done.

gjkim42 commented:
/retest

gjkim42 commented:
@thockin 
Could you PTAL? Thanks!

aojea commented on pkg/kubelet/network/dns/dns.go:
```diff
@@ -99,6 +100,14 @@ func omitDuplicates(strs []string) []string {
+		if len(search) > utilvalidation.DNS1123SubdomainMaxLength {
+			composedSearch = composedSearch[:i]
+			limitsExceeded = true
+			break
```
this should be a continue if we want to have all valid "searches" https://github.com/kubernetes/kubernetes/pull/100651/commits/a974fb5f8883506d1456907d23d26ca0195120ce#r635548677

aojea replied:
no need to continue since there is no more code after, just removing it

gjkim42 replied:
I think you are commenting on outdated code.

aojea replied:
yeah, I just realised, sorry

aojea commented on pkg/kubelet/network/dns/dns.go:
```diff
@@ -99,6 +100,14 @@ func omitDuplicates(strs []string) []string {
 func (c *Configurer) formDNSSearchFitsLimits(composedSearch []string, pod *v1.Pod) []string {
 	limitsExceeded := false
 
+	for i, search := range composedSearch {
```
a comment explaining what we are doing, something like
```
// In glibc 2.28, DNS resolving causes abort() if there is a search path exceeding 255 characters
// so we filter them out
```

aojea replied:
it is comment in an outdated commit, but I think it applies to the new code :)

gjkim42 replied:
Thanks. I will add the comment :)

aojea commented on pkg/kubelet/network/dns/dns_test.go:
```diff
@@ -121,43 +145,93 @@ func TestFormDNSSearchFitsLimits(t *testing.T) {
+			events:       []string{},
+		},
+
+		{
+			desc:         "invalid: longer than 256 characters in search path list",
+			hostNames:    []string{"testNS.svc.TEST", "svc.TEST", "TEST", "AAA", strings.Repeat("B", 256), "BBB"},
+			resultSearch: []string{"testNS.svc.TEST", "svc.TEST", "TEST", "AAA", "BBB"},
+			events:       []string{"Search Line limits were exceeded, some search paths have been omitted, the applied search line is: testNS.svc.TEST svc.TEST TEST AAA BBB"},
+		},
```
can we add one with the offender in the middle of the list instead of being the last one?

gjkim42 replied:
I think the case already has a long search path in the middle of the list.
Let me know if I miss something.

aojea replied:
```
hostNames:    []string{"testNS.svc.TEST", strings.Repeat("B", 256), "BBB","svc.TEST", "TEST", "AAA", },			resultSearch: []string{"testNS.svc.TEST", "svc.TEST", "TEST", "AAA", "BBB"},
```

gjkim42 replied:
I am not sure what you said...
I think `strings.Repeat("B", 256)` should be filtered out and it already is the one in the middle of the list(before the "BBB").

aojea replied:
ahhh sorry, I totally missed that last "BBB", my apologies

aojea commented on pkg/api/pod/util.go:
```diff
@@ -390,6 +390,37 @@ func usesIndivisibleHugePagesValues(podSpec *api.PodSpec) bool {
 
+// isSameExpandedDNSConfig returns true if the oldPodSpec already had
+// ExpandedDNSConfig and podSpec has the same DNSConfig
+func isSameExpandedDNSConfig(podSpec, oldPodSpec *api.PodSpec) bool {
```
should we add an unit test for this function to protect this code?

gjkim42 replied:
Added a unit test.
I think this function can be removed after the alpha stage(?).
Until then, it would be nice to test the functionality.

aojea commented:
some comments but I think that is ready for final review
/hold cancel

gjkim42 committed:
kubelet: Validate the length of the DNS search path

gjkim42 committed:
Add feature gate ExpandedDNSConfig

ExpandedDNSConfig allows kubernetes to have expanded DNS(Domain Name
System) configuration

gjkim42 committed:
AllowExpandedDNSConfig if haveSameExpandedDNSConfig(newPod, oldPod)

gjkim42 commented:
/retest

gjkim42 commented:
/retest

gjkim42 commented:
/retest

aojea commented:
/lgtm
/assign @thockin

thockin commented on pkg/kubelet/network/dns/dns.go:
```diff
@@ -99,19 +102,37 @@ func omitDuplicates(strs []string) []string {
+		composedSearch[l] = search
+		l++
+	}
+	composedSearch = composedSearch[:l]
```
I am missing what this line is doing?

gjkim42 replied:
L117-L126 filters out search paths that exceed 255 characters for DNS resolvers to not explode.

gjkim42 replied:
https://github.com/kubernetes/kubernetes/pull/100651#discussion_r635548677

thockin replied:
I see, I missed that it was an in-place filter.

thockin reviewed (commented):
One question

thockin reviewed (commented):
Thanks!

/lgtm
/approve

gjkim42 commented:
Rebased
Need LGTM :)

gjkim42 commented:
@thockin @aojea 
Could you re-LGTM on this?

thockin reviewed (commented):
Thanks!

/lgtm
/approve
